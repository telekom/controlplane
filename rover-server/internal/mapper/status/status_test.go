// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package status

import (
	"github.com/gkampitakis/go-snaps/match"
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/telekom/controlplane/common/pkg/condition"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Rover Status Mapper", func() {
	Context("MapRoverStatus", func() {
		It("must map rover status correctly", func() {
			status := MapRoverStatus(ctx, rover)

			Expect(status).ToNot(BeNil())

			snaps.MatchJSON(GinkgoT(), status, match.Any("time"))
		})

	})

	Context("MapRoverResponse", func() {
		It("must map rover response correctly", func() {
			response, err := MapRoverResponse(ctx, rover)

			Expect(response).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), response)

			Expect(err).To(BeNil())
		})

		It("must return an error if the input rover is nil", func() {
			response, err := MapRoverResponse(ctx, nil)

			Expect(response).ToNot(BeNil())

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("input rover is nil"))
		})
	})

	Context("GetOverallStatus", func() {
		It("must get overall status correctly", func() {
			status := GetOverallStatus(rover.GetConditions())

			Expect(status).ToNot(BeNil())
			Expect(status).To(Equal(api.OverallStatusBlocked))
		})

		It("must get overall status correctly if the input is nil", func() {
			status := GetOverallStatus(nil)

			Expect(status).ToNot(BeNil())
			Expect(status).To(Equal(api.OverallStatusNone))
		})
	})

	Context("FillStateInfo", func() {
		It("if no processing condition exists status will be none and warnings are created", func() {
			var conditions []metav1.Condition
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if no ready condition exists status will be none and warnings are created", func() {
			conditions := []metav1.Condition{
				{
					Type:   condition.ConditionTypeProcessing,
					Status: metav1.ConditionTrue,
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if processing condition status will be blocked and processing", func() {
			conditions := []metav1.Condition{
				{
					Type:   condition.ConditionTypeProcessing,
					Status: metav1.ConditionTrue,
				},
				{
					Type:   condition.ConditionTypeReady,
					Status: metav1.ConditionFalse,
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if processing condition is blocked status will be blocked and warning", func() {
			conditions := []metav1.Condition{
				{
					Type:    condition.ConditionTypeProcessing,
					Reason:  "Blocked",
					Message: "Blocked due to dependency",
				},
				{
					Type:   condition.ConditionTypeReady,
					Status: metav1.ConditionFalse,
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if processing done and ready true status will be complete", func() {
			conditions := []metav1.Condition{
				{
					Type:   condition.ConditionTypeProcessing,
					Reason: "Done",
				},
				{
					Type:   condition.ConditionTypeReady,
					Status: metav1.ConditionTrue,
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if processing done and ready false status will be blocked and warning", func() {
			conditions := []metav1.Condition{
				{
					Type:   condition.ConditionTypeProcessing,
					Reason: "Done",
				},
				{
					Type:    condition.ConditionTypeReady,
					Status:  metav1.ConditionFalse,
					Message: "Not ready yet",
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

		It("if processing failed status will be failed and error", func() {
			conditions := []metav1.Condition{
				{
					Type:    condition.ConditionTypeProcessing,
					Reason:  "Failed",
					Message: "Processing failed",
				},
				{
					Type:   condition.ConditionTypeReady,
					Status: metav1.ConditionFalse,
				},
			}
			status := &api.Status{}

			fillStateInfo(conditions, status)

			Expect(status).ToNot(BeNil())
			snaps.MatchJSON(GinkgoT(), status)
		})

	})
})
