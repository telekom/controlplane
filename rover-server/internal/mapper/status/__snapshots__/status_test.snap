
[Rover Status Mapper MapRoverStatus must map rover status correctly - 1]
{
 "errors": [
  {
   "cause": "Application: rover-local-sub, Cause: NoApproval",
   "message": "Approval is either rejected or suspended"
  }
 ],
 "processingState": "done",
 "state": "blocked",
 "time": "<Any value>",
 "warnings": [
  {
   "message": "Atleast one sub-resource is being processed"
  }
 ]
}
---

[Rover Status Mapper MapRoverResponse must map rover response correctly - 1]
{
 "errors": [
  {
   "cause": "NoApproval",
   "context": "Application: rover-local-sub",
   "details": "Condition: Ready, Status: False, Message: Approval is either rejected or suspended",
   "message": "Approval is either rejected or suspended",
   "resource": {
    "apiVersion": "rover.cp.ei.telekom.de/v1",
    "kind": "Rover",
    "name": "rover-local-sub",
    "namespace": "poc--eni--hyperion"
   }
  }
 ],
 "overallStatus": "blocked",
 "processingState": "done",
 "state": "blocked"
}
---

[Rover Status Mapper FillStateInfo if no processing condition exists status will be none and warnings are created - 1]
{
 "processingState": "none",
 "state": "none",
 "time": "",
 "warnings": [
  {
   "message": "Processing condition not found"
  }
 ]
}
---

[Rover Status Mapper FillStateInfo if no ready condition exists status will be none and warnings are created - 1]
{
 "processingState": "none",
 "state": "none",
 "time": "",
 "warnings": [
  {
   "message": "Ready condition not found"
  }
 ]
}
---

[Rover Status Mapper FillStateInfo if processing condition status will be blocked and processing - 1]
{
 "processingState": "processing",
 "state": "blocked",
 "time": ""
}
---

[Rover Status Mapper FillStateInfo if processing condition is blocked status will be blocked and warning - 1]
{
 "processingState": "done",
 "state": "blocked",
 "time": "",
 "warnings": [
  {
   "message": "Blocked due to dependency"
  }
 ]
}
---

[Rover Status Mapper FillStateInfo if processing done and ready true status will be complete - 1]
{
 "processingState": "done",
 "state": "complete",
 "time": ""
}
---

[Rover Status Mapper FillStateInfo if processing done and ready false status will be blocked and warning - 1]
{
 "processingState": "done",
 "state": "blocked",
 "time": "",
 "warnings": [
  {
   "message": "Not ready yet"
  }
 ]
}
---

[Rover Status Mapper FillStateInfo if processing failed status will be failed and error - 1]
{
 "errors": [
  {
   "message": "Processing failed"
  }
 ],
 "processingState": "failed",
 "state": "",
 "time": ""
}
---
